Sources/ProcedureKit/ProcedureQueue.swift

```diff
@@ -6,103 +6,148 @@
 
 import Foundation
 
-public protocol OperationQueueDelegate: class {
+/**
+ A protocol which the `ProcedureQueue`'s delegate must conform to. The delegate is informed
+ when the queue is about to add an operation/procedure, and when they finish. Because it is a
+ delegate protocol, conforming types must be classes, as the queue weakly owns it.
+ 
+ There are slight differences depending on whether an Operation subclass or a Procedure subclass
+ is being added to a ProcedureQueue.
+ 
+ For an Operation subclass, the delegate callbacks are:
+ 
+    - procedureQueue(_:, willAddOperation:, context:)
+    - procedureQueue(_:, didAddOperation:, context:)
+    - procedureQueue(_:, didFinishOperation:)
+ 
+ If you get the first, you are guaranteed* to get the other two callbacks once the Operation
+ finishes (*unless you later modify the completionBlock - which is a bad idea - or your Operation
+ never finishes - which is also a bad idea).
+ 
+ For a Procedure subclass, the delegate callbacks are:
+ 
+    - procedureQueue(_:, willAddProcedure:, context:)
+    - procedureQueue(_:, didAddProcedure:, context:)
+    - procedureQueue(_:, willFinishProcedure:)
+    - procedureQueue(_:, didFinishProcedure:)
+ 
+ You will (eventually) receive all 4 delegate callbacks for a Procedure (assuming it finishes).
+ 
+ The `context` parameter provided to the will/didAdd delegate callbacks is the same `context`
+ parameter that was passed into the call to `ProcedureQueue.add(operation:context:)`.
+ 
+ For both adding and finishing, the "will" delegate will be called before the respective "did" delegate
+ method. However, you should be aware of the following additional guidelines:
+ 
+    - `willAddOperation` / `willAddProcedure` is always guaranteed to be called before any other delegate
+      methods for an Operation / Procedure instance
+    - `willFinishProcedure` is always guaranteed to be called before `didFinishProcedure` for a
+      Procedure instance
+    - No other ordering (or non-concurrency) is guaranteed
+ 
+ Delegate callbacks may occur concurrently and on any thread.
+ 
+ Examples:
+    - It is possible for `didAddOperation` / `didAddProcedure` to be called concurrently with
+    `didFinishOperation` / `will/didFinishProcedure` for an Operation / Procedure instance.
+    - It is possible for `didAddOperation` to be called concurrently for two different Operation
+    instances added to a ProcedureQueue simultaneously. (i.e. Once for each Operation.)
+ 
+ */
+public protocol ProcedureQueueDelegate: class {
```

Comment:
Until any Swift language change on this is finalized, I think it's reasonable to match the current Swift documentation and the pattern used elsewhere in ProcedureKit (which is `class`).

(I do see the argument for changing the language, though.)